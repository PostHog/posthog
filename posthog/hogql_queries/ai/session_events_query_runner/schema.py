"""
Schema definitions for SessionBatchEventsQuery extending EventsQuery.

This approach leverages PostHog's existing EventsQuery infrastructure while adding
multi-session capabilities for session summary workflows.
"""

from __future__ import annotations

from typing import Any, Optional

from posthog.session_recordings.constants import EXTRA_SUMMARY_EVENT_FIELDS
from posthog.schema import SessionBatchEventsQuery
from posthog.session_recordings.queries.session_replay_events import DEFAULT_EVENT_FIELDS


# Type alias for convenience
SessionEventsResults = dict[str, list[list[Any]]]  # session_id -> events mapping


def create_session_batch_events_query(
    session_ids: list[str],
    select: Optional[list[str]] = None,
    events_to_ignore: Optional[list[str]] = None,
    after: Optional[str] = None,
    before: Optional[str] = None,
    max_total_events: Optional[int] = None,
    offset: Optional[int] = None,
    include_session_id: bool = True,
    **kwargs: Any,
) -> SessionBatchEventsQuery:
    """Create query for getting events for multiple sessions"""

    # Default field selection for session summaries
    if select is None:
        select = DEFAULT_EVENT_FIELDS + EXTRA_SUMMARY_EVENT_FIELDS

    # Ensure $session_id is included for grouping
    if include_session_id and "properties.$session_id" not in select:
        select.append("properties.$session_id")

    # Build WHERE clauses for session filtering and event exclusion
    where_clauses = [f"properties.$session_id IN {session_ids}"]  # Filter by session IDs by default

    # Exclude unwanted events (default to ignoring feature flag calls)
    if events_to_ignore is None:
        events_to_ignore = ["$feature_flag_called"]
    if events_to_ignore:
        event_list = "', '".join(events_to_ignore)
        where_clauses.append(f"event NOT IN ('{event_list}')")

    # Combine with any existing where clauses, if provided
    existing_where = kwargs.get("where", [])
    if isinstance(existing_where, list):
        where_clauses.extend(existing_where)

    # Set defaults for session batch queries
    query_params = {
        "select": select,
        "session_ids": session_ids,
        "where": where_clauses,
        "orderBy": ["properties.$session_id", "timestamp ASC"],  # Group by session, then chronological
        "limit": max_total_events,
        "offset": offset,
        "group_by_session": True,
        "after": after,
        "before": before,
        **kwargs,  # Allow overriding defaults
    }
    return SessionBatchEventsQuery(**query_params)
