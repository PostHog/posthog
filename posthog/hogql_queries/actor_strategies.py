from typing import Dict, List, cast, Literal, Optional

from django.db.models import Prefetch

from posthog.hogql import ast
from posthog.hogql.property import property_to_expr
from posthog.hogql_queries.insights.paginators import HogQLHasMorePaginator
from posthog.models import Team, Person, Group
from posthog.schema import PersonsQuery


class ActorStrategy:
    field: str
    origin: str
    origin_id: str

    def __init__(self, team: Team, query: PersonsQuery, paginator: HogQLHasMorePaginator):
        self.team = team
        self.paginator = paginator
        self.query = query

    def get_actors(self, actor_ids) -> Dict[str, Dict]:
        raise NotImplementedError()

    def input_columns(self) -> List[str]:
        raise NotImplementedError()

    def filter_conditions(self) -> List[ast.Expr]:
        return []

    def order_by(self) -> Optional[List[ast.OrderExpr]]:
        return None


class PersonStrategy(ActorStrategy):
    field = "person"
    origin = "persons"
    origin_id = "id"

    def get_actors(self, actor_ids) -> Dict[str, Dict]:
        return {
            str(p.uuid): {
                "id": p.uuid,
                **{field: getattr(p, field) for field in ("distinct_ids", "properties", "created_at", "is_identified")},
            }
            for p in Person.objects.filter(
                team_id=self.team.pk, persondistinctid__team_id=self.team.pk, uuid__in=actor_ids
            )
            .prefetch_related(Prefetch("persondistinctid_set", to_attr="distinct_ids_cache"))
            .iterator(chunk_size=self.paginator.limit)
        }

    def input_columns(self) -> List[str]:
        return ["person", "id", "created_at", "person.$delete"]

    def filter_conditions(self) -> List[ast.Expr]:
        where_exprs: List[ast.Expr] = []

        if self.query.properties:
            where_exprs.append(property_to_expr(self.query.properties, self.team, scope="person"))

        if self.query.fixedProperties:
            where_exprs.append(property_to_expr(self.query.fixedProperties, self.team, scope="person"))

        if self.query.search is not None and self.query.search != "":
            where_exprs.append(
                ast.Or(
                    exprs=[
                        ast.CompareOperation(
                            op=ast.CompareOperationOp.ILike,
                            left=ast.Field(chain=["properties", "email"]),
                            right=ast.Constant(value=f"%{self.query.search}%"),
                        ),
                        ast.CompareOperation(
                            op=ast.CompareOperationOp.ILike,
                            left=ast.Field(chain=["properties", "name"]),
                            right=ast.Constant(value=f"%{self.query.search}%"),
                        ),
                        ast.CompareOperation(
                            op=ast.CompareOperationOp.ILike,
                            left=ast.Call(name="toString", args=[ast.Field(chain=["id"])]),
                            right=ast.Constant(value=f"%{self.query.search}%"),
                        ),
                        ast.CompareOperation(
                            op=ast.CompareOperationOp.ILike,
                            left=ast.Field(chain=["pdi", "distinct_id"]),
                            right=ast.Constant(value=f"%{self.query.search}%"),
                        ),
                    ]
                )
            )
        return where_exprs

    def order_by(self) -> Optional[List[ast.OrderExpr]]:
        if self.query.orderBy not in [["person"], ["person DESC"], ["person ASC"]]:
            return None

        order_property = (
            "email" if self.team.person_display_name_properties is None else self.team.person_display_name_properties[0]
        )
        return [
            ast.OrderExpr(
                expr=ast.Field(chain=["properties", order_property]),
                order=cast(
                    Literal["ASC", "DESC"],
                    "DESC" if self.query.orderBy[0] == "person DESC" else "ASC",
                ),
            )
        ]


class GroupStrategy(ActorStrategy):
    field = "group"
    origin = "groups"
    origin_id = "key"

    def __init__(self, group_type_index: int, **kwargs):
        self.group_type_index = group_type_index
        super().__init__(**kwargs)

    def get_actors(self, actor_ids) -> Dict[str, Dict]:
        return {
            str(p["group_key"]): {
                "id": p["group_key"],
                "type": "group",
                "properties": p["group_properties"],  # TODO: Legacy for frontend
                **p,
            }
            for p in Group.objects.filter(
                team_id=self.team.pk, group_type_index=self.group_type_index, group_key__in=actor_ids
            )
            .values("group_key", "group_type_index", "created_at", "group_properties")
            .iterator(chunk_size=self.paginator.limit)
        }

    def input_columns(self) -> List[str]:
        return ["group"]

    def filter_conditions(self) -> List[ast.Expr]:
        where_exprs: List[ast.Expr] = []

        if self.query.search is not None and self.query.search != "":
            where_exprs.append(
                ast.Or(
                    exprs=[
                        ast.CompareOperation(
                            op=ast.CompareOperationOp.ILike,
                            left=ast.Field(chain=["properties", "name"]),
                            right=ast.Constant(value=f"%{self.query.search}%"),
                        ),
                        ast.CompareOperation(
                            op=ast.CompareOperationOp.ILike,
                            left=ast.Call(name="toString", args=[ast.Field(chain=["key"])]),
                            right=ast.Constant(value=f"%{self.query.search}%"),
                        ),
                    ]
                )
            )

        return where_exprs

    def order_by(self) -> Optional[List[ast.OrderExpr]]:
        if self.query.orderBy not in [["group"], ["group DESC"], ["group ASC"]]:
            return None

        order_property = "name"
        return [
            ast.OrderExpr(
                expr=ast.Field(chain=["properties", order_property]),
                order=cast(
                    Literal["ASC", "DESC"],
                    "DESC" if self.query.orderBy[0] == "group DESC" else "ASC",
                ),
            )
        ]
