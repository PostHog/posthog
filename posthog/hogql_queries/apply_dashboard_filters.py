from posthog.hogql_queries.query_runner import get_query_runner
from posthog.models import Team
from posthog.schema import DashboardFilter


# Apply the filters from the django-style Dashboard object
def apply_dashboard_filters(query: dict, filters: dict, team: Team) -> dict:
    kind = query.get("kind", None)

    if kind == "DataTableNode":
        source = apply_dashboard_filters(query["source"], filters, team)
        return {**query, "source": source}

    try:
        query_runner = get_query_runner(query, team)
    except ValueError:
        return query
    try:
        return query_runner.apply_dashboard_filters(DashboardFilter(**filters)).dict()
    except NotImplementedError:
        # TODO when we implement apply_dashboard_filters on more query runners, we can remove the try/catch
        return query
