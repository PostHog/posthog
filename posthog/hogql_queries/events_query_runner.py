import json
from datetime import timedelta
from typing import Dict, List, Optional, Any

from dateutil.parser import isoparse
from django.db.models import Prefetch
from django.utils.timezone import now

from posthog.api.element import ElementSerializer
from posthog.api.utils import get_pk_or_uuid
from posthog.clickhouse.client.connection import Workload
from posthog.hogql import ast
from posthog.hogql.parser import parse_expr, parse_order_expr
from posthog.hogql.property import action_to_expr, has_aggregation, property_to_expr
from posthog.hogql.query import execute_hogql_query
from posthog.hogql.timings import HogQLTimings
from posthog.hogql_queries.query_runner import QueryRunner
from posthog.models import Action, Person, Team
from posthog.models.element import chain_to_elements
from posthog.models.person.util import get_persons_by_distinct_ids
from posthog.schema import EventsQuery, EventsQueryResponse
from posthog.utils import relative_date_parse

# Allow-listed fields returned when you select "*" from events. Person and group fields will be nested later.
SELECT_STAR_FROM_EVENTS_FIELDS = [
    "uuid",
    "event",
    "properties",
    "timestamp",
    "team_id",
    "distinct_id",
    "elements_chain",
    "created_at",
]


class EventsQueryRunner(QueryRunner):
    query: EventsQuery
    query_type = EventsQuery

    def __init__(
        self,
        query: EventsQuery | Dict[str, Any],
        team: Team,
        timings: Optional[HogQLTimings] = None,
        in_export_context: Optional[bool] = False,
    ):
        super().__init__(query, team, timings, in_export_context)
        if isinstance(query, EventsQuery):
            self.query = query
        else:
            self.query = EventsQuery.model_validate(query)

    def to_query(self) -> ast.SelectQuery:
        # Note: This code is inefficient and problematic, see https://github.com/PostHog/posthog/issues/13485 for details.
        if self.timings is None:
            self.timings = HogQLTimings()

        with self.timings.measure("build_ast"):
            # limit & offset
            offset = 0 if self.query.offset is None else self.query.offset

            # columns & group_by
            with self.timings.measure("columns"):
                select_input: List[str] = []
                person_indices: List[int] = []
                for index, col in enumerate(self.select_input_raw()):
                    # Selecting a "*" expands the list of columns, resulting in a table that's not what we asked for.
                    # Instead, ask for a tuple with all the columns we want. Later transform this back into a dict.
                    if col == "*":
                        select_input.append(f"tuple({', '.join(SELECT_STAR_FROM_EVENTS_FIELDS)})")
                    elif col.split("--")[0].strip() == "person":
                        # This will be expanded into a followup query
                        select_input.append("distinct_id")
                        person_indices.append(index)
                    else:
                        select_input.append(col)
                select: List[ast.Expr] = [parse_expr(column, timings=self.timings) for column in select_input]

            with self.timings.measure("aggregations"):
                group_by: List[ast.Expr] = [column for column in select if not has_aggregation(column)]
                aggregations: List[ast.Expr] = [column for column in select if has_aggregation(column)]
                has_any_aggregation = len(aggregations) > 0

            # filters
            with self.timings.measure("filters"):
                with self.timings.measure("where"):
                    where_input = self.query.where or []
                    where_exprs = [parse_expr(expr, timings=self.timings) for expr in where_input]
                if self.query.properties:
                    with self.timings.measure("properties"):
                        where_exprs.extend(property_to_expr(property, self.team) for property in self.query.properties)
                if self.query.fixedProperties:
                    with self.timings.measure("fixed_properties"):
                        where_exprs.extend(
                            property_to_expr(property, self.team) for property in self.query.fixedProperties
                        )
                if self.query.event:
                    with self.timings.measure("event"):
                        where_exprs.append(
                            parse_expr(
                                "event = {event}",
                                {"event": ast.Constant(value=self.query.event)},
                                timings=self.timings,
                            )
                        )
                if self.query.actionId:
                    with self.timings.measure("action_id"):
                        try:
                            action = Action.objects.get(pk=self.query.actionId, team_id=self.team.pk)
                        except Action.DoesNotExist:
                            raise Exception("Action does not exist")
                        if action.steps.count() == 0:
                            raise Exception("Action does not have any match groups")
                        where_exprs.append(action_to_expr(action))
                if self.query.personId:
                    with self.timings.measure("person_id"):
                        person: Optional[Person] = get_pk_or_uuid(Person.objects.all(), self.query.personId).first()
                        distinct_ids = person.distinct_ids if person is not None else []
                        ids_list = list(map(str, distinct_ids))
                        where_exprs.append(
                            parse_expr(
                                "distinct_id in {list}",
                                {"list": ast.Constant(value=ids_list)},
                                timings=self.timings,
                            )
                        )

            with self.timings.measure("timestamps"):
                # prevent accidentally future events from being visible by default
                before = self.query.before or (now() + timedelta(seconds=5)).isoformat()
                try:
                    parsed_date = isoparse(before)
                except ValueError:
                    parsed_date = relative_date_parse(before, self.team.timezone_info)
                where_exprs.append(
                    parse_expr(
                        "timestamp < {timestamp}",
                        {"timestamp": ast.Constant(value=parsed_date)},
                        timings=self.timings,
                    )
                )

                # limit to the last 24h by default
                after = self.query.after or "-24h"
                if after != "all":
                    try:
                        parsed_date = isoparse(after)
                    except ValueError:
                        parsed_date = relative_date_parse(after, self.team.timezone_info)
                    where_exprs.append(
                        parse_expr(
                            "timestamp > {timestamp}",
                            {"timestamp": ast.Constant(value=parsed_date)},
                            timings=self.timings,
                        )
                    )

            # where & having
            with self.timings.measure("where"):
                where_list = [expr for expr in where_exprs if not has_aggregation(expr)]
                where = ast.And(exprs=where_list) if len(where_list) > 0 else None
                having_list = [expr for expr in where_exprs if has_aggregation(expr)]
                having = ast.And(exprs=having_list) if len(having_list) > 0 else None

            # order by
            with self.timings.measure("order"):
                if self.query.orderBy is not None:
                    order_by = [parse_order_expr(column, timings=self.timings) for column in self.query.orderBy]
                elif "count()" in select_input:
                    order_by = [ast.OrderExpr(expr=parse_expr("count()"), order="DESC")]
                elif len(aggregations) > 0:
                    order_by = [ast.OrderExpr(expr=aggregations[0], order="DESC")]
                elif "timestamp" in select_input:
                    order_by = [ast.OrderExpr(expr=ast.Field(chain=["timestamp"]), order="DESC")]
                elif len(select) > 0:
                    order_by = [ast.OrderExpr(expr=select[0], order="ASC")]
                else:
                    order_by = []

            with self.timings.measure("select"):
                stmt = ast.SelectQuery(
                    select=select,
                    select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
                    where=where,
                    having=having,
                    group_by=group_by if has_any_aggregation else None,
                    order_by=order_by,
                    limit=ast.Constant(value=self.limit()),
                    offset=ast.Constant(value=offset),
                )
                return stmt

    def calculate(self) -> EventsQueryResponse:
        query_result = execute_hogql_query(
            query=self.to_query(),
            team=self.team,
            workload=Workload.ONLINE,
            query_type="EventsQuery",
            timings=self.timings,
            in_export_context=self.in_export_context,
        )

        # Convert star field from tuple to dict in each result
        if "*" in self.select_input_raw():
            with self.timings.measure("expand_asterisk"):
                star_idx = self.select_input_raw().index("*")
                for index, result in enumerate(query_result.results):
                    query_result.results[index] = list(result)
                    select = result[star_idx]
                    new_result = dict(zip(SELECT_STAR_FROM_EVENTS_FIELDS, select))
                    new_result["properties"] = json.loads(new_result["properties"])
                    if new_result["elements_chain"]:
                        new_result["elements"] = ElementSerializer(
                            chain_to_elements(new_result["elements_chain"]), many=True
                        ).data
                    query_result.results[index][star_idx] = new_result

        person_indices: List[int] = []
        for index, col in enumerate(self.select_input_raw()):
            if col.split("--")[0].strip() == "person":
                person_indices.append(index)

        if len(person_indices) > 0 and len(query_result.results) > 0:
            with self.timings.measure("person_column_extra_query"):
                # Make a query into postgres to fetch person
                person_idx = person_indices[0]
                distinct_ids = list(set(event[person_idx] for event in query_result.results))
                persons = get_persons_by_distinct_ids(self.team.pk, distinct_ids)
                persons = persons.prefetch_related(Prefetch("persondistinctid_set", to_attr="distinct_ids_cache"))
                distinct_to_person: Dict[str, Person] = {}
                for person in persons:
                    if person:
                        for person_distinct_id in person.distinct_ids:
                            distinct_to_person[person_distinct_id] = person

                # Loop over all columns in case there is more than one "person" column
                for column_index in person_indices:
                    for index, result in enumerate(query_result.results):
                        distinct_id: str = result[column_index]
                        query_result.results[index] = list(result)
                        if distinct_to_person.get(distinct_id):
                            person = distinct_to_person[distinct_id]
                            query_result.results[index][column_index] = {
                                "uuid": person.uuid,
                                "created_at": person.created_at,
                                "properties": person.properties or {},
                                "distinct_id": distinct_id,
                            }
                        else:
                            query_result.results[index][column_index] = {
                                "distinct_id": distinct_id,
                            }

        received_extra_row = len(query_result.results) == self.limit()  # limit was +=1'd above
        return EventsQueryResponse(
            results=query_result.results[: self.limit() - 1] if received_extra_row else query_result.results,
            columns=self.select_input_raw(),
            types=[type for _, type in query_result.types],
            hasMore=received_extra_row,
            timings=self.timings.to_list(),
            hogql=query_result.hogql,
        )

    def select_input_raw(self) -> List[str]:
        return ["*"] if len(self.query.select) == 0 else self.query.select

    def limit(self) -> int:
        # importing locally so we could override in a test
        from posthog.hogql.constants import (
            DEFAULT_RETURNED_ROWS,
            MAX_SELECT_RETURNED_ROWS,
        )

        # adding +1 to the limit to check if there's a "next page" after the requested results
        return (
            min(
                MAX_SELECT_RETURNED_ROWS,
                (MAX_SELECT_RETURNED_ROWS if self.in_export_context else DEFAULT_RETURNED_ROWS)
                if self.query.limit is None
                else self.query.limit,
            )
            + 1
        )

    def _is_stale(self, cached_result_package):
        return True

    def _refresh_frequency(self):
        return timedelta(minutes=1)
