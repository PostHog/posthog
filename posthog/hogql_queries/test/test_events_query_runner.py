from typing import Any, cast

from freezegun import freeze_time
from datetime import datetime
from posthog.hogql import ast
from posthog.hogql.ast import CompareOperationOp
from posthog.hogql_queries.events_query_runner import EventsQueryRunner
from posthog.models import Person, Team
from posthog.models.organization import Organization
from posthog.schema import (
    CachedEventsQueryResponse,
    EventsQuery,
    EventPropertyFilter,
    PropertyOperator,
)
from posthog.test.base import (
    APIBaseTest,
    ClickhouseTestMixin,
    _create_event,
    _create_person,
    also_test_with_different_timezones,
    flush_persons_and_events,
    snapshot_clickhouse_queries,
)


class TestEventsQueryRunner(ClickhouseTestMixin, APIBaseTest):
    maxDiff = None

    def _create_events(self, data: list[tuple[str, str, Any]], event="$pageview"):
        person_result = []
        distinct_ids_handled = set()
        for distinct_id, timestamp, event_properties in data:
            with freeze_time(timestamp):
                if distinct_id not in distinct_ids_handled:
                    person_result.append(
                        _create_person(
                            team_id=self.team.pk,
                            distinct_ids=[distinct_id],
                            properties={
                                "name": distinct_id,
                            },
                        )
                    )
                    distinct_ids_handled.add(distinct_id)
                _create_event(
                    team=self.team,
                    event=event,
                    distinct_id=distinct_id,
                    timestamp=timestamp,
                    properties=event_properties,
                )
        return person_result

    def _create_boolean_field_test_events(self):
        self._create_events(
            data=[
                (
                    "p_true",
                    "2020-01-11T12:00:01Z",
                    {"boolean_field": True},
                ),
                (
                    "p_false",
                    "2020-01-11T12:00:02Z",
                    {"boolean_field": False},
                ),
                (
                    "p_notset",
                    "2020-01-11T12:00:04Z",
                    {},
                ),
                (
                    "p_null",
                    "2020-01-11T12:00:04Z",
                    {"boolean_field": None},
                ),
            ]
        )

    def _run_boolean_field_query(self, filter: EventPropertyFilter):
        with freeze_time("2020-01-11T12:01:00"):
            query = EventsQuery(
                after="-24h",
                event="$pageview",
                kind="EventsQuery",
                orderBy=["timestamp ASC"],
                select=["*"],
                properties=[filter],
            )

            runner = EventsQueryRunner(query=query, team=self.team)
            response = runner.run()
            assert isinstance(response, CachedEventsQueryResponse)
            results = response.results
            return results

    def test_is_not_set_boolean(self):
        # see https://github.com/PostHog/posthog/issues/18030
        self._create_boolean_field_test_events()
        results = self._run_boolean_field_query(
            EventPropertyFilter(
                type="event",
                key="boolean_field",
                operator=PropertyOperator.IS_NOT_SET,
                value=PropertyOperator.IS_NOT_SET,
            )
        )

        self.assertEqual({"p_notset", "p_null"}, {row[0]["distinct_id"] for row in results})

    def test_is_set_boolean(self):
        self._create_boolean_field_test_events()

        results = self._run_boolean_field_query(
            EventPropertyFilter(
                type="event",
                key="boolean_field",
                operator=PropertyOperator.IS_SET,
                value=PropertyOperator.IS_SET,
            )
        )

        self.assertEqual({"p_true", "p_false"}, {row[0]["distinct_id"] for row in results})

    def test_person_id_expands_to_distinct_ids(self):
        _create_person(
            team_id=self.team.pk,
            distinct_ids=["id1", "id2"],
        )
        flush_persons_and_events()
        person = Person.objects.first()
        query = EventsQuery(kind="EventsQuery", select=["*"], personId=str(person.pk))  # type: ignore

        # matching team
        query_ast = EventsQueryRunner(query=query, team=self.team).to_query()
        where_expr = cast(ast.CompareOperation, cast(ast.And, query_ast.where).exprs[0])
        right_expr = cast(ast.Tuple, where_expr.right)
        self.assertEqual(
            [cast(ast.Constant, cast(ast.Call, x).args[0]).value for x in right_expr.exprs], ["id1", "id2"]
        )

        # another team
        another_team = Team.objects.create(organization=Organization.objects.create())
        query_ast = EventsQueryRunner(query=query, team=another_team).to_query()
        where_expr = cast(ast.CompareOperation, cast(ast.And, query_ast.where).exprs[0])
        right_expr = cast(ast.Tuple, where_expr.right)
        self.assertEqual(right_expr.exprs, [])

    def test_test_account_filters(self):
        self.team.test_account_filters = [
            {
                "key": "email",
                "type": "person",
                "value": "posthog.com",
                "operator": "not_icontains",
            }
        ]
        self.team.save()
        query = EventsQuery(kind="EventsQuery", select=["*"], filterTestAccounts=True)
        query_ast = EventsQueryRunner(query=query, team=self.team).to_query()
        where_expr = cast(ast.CompareOperation, cast(ast.And, query_ast.where).exprs[0])
        right_expr = cast(ast.Constant, where_expr.right)
        self.assertEqual(right_expr.value, "%posthog.com%")
        self.assertEqual(where_expr.op, CompareOperationOp.NotILike)

    def test_big_int(self):
        BIG_INT = 2**159 - 24
        self._create_events(
            data=[
                (
                    "p_null",
                    "2020-01-11T12:00:04Z",
                    {"boolean_field": None, "bigInt": BIG_INT},
                ),
            ]
        )

        flush_persons_and_events()

        with freeze_time("2020-01-11T12:01:00"):
            query = EventsQuery(
                after="-24h",
                event="$pageview",
                kind="EventsQuery",
                orderBy=["timestamp ASC"],
                select=["*"],
            )

            runner = EventsQueryRunner(query=query, team=self.team)
            response = runner.run()
            assert isinstance(response, CachedEventsQueryResponse)
            assert response.results[0][0]["properties"]["bigInt"] == float(BIG_INT)

    def test_escaped_single_quotes_in_where_clause(self):
        SINGLE_QUOTE = "I'm a string with a ' in it"
        DOUBLE_QUOTE = 'I"m a string with a " in it'
        self._create_events(
            data=[
                (
                    "p_null",
                    "2020-01-11T12:00:04Z",
                    {"boolean_field": None, "arr_field": [SINGLE_QUOTE]},
                ),
                (
                    "p_one",
                    "2020-01-11T12:00:14Z",
                    {"boolean_field": None, "arr_field": [DOUBLE_QUOTE]},
                ),
            ]
        )

        flush_persons_and_events()

        with freeze_time("2020-01-11T12:01:00"):
            query = EventsQuery(
                after="-24h",
                event="$pageview",
                kind="EventsQuery",
                where=[
                    "has(JSONExtract(ifNull(properties.arr_field,'[]'),'Array(String)'), 'I\\'m a string with a \\' in it')"
                ],
                orderBy=["timestamp ASC"],
                select=["*"],
            )

            runner = EventsQueryRunner(query=query, team=self.team)
            response = runner.run()
            assert isinstance(response, CachedEventsQueryResponse)
            assert len(response.results) == 1
            assert response.results[0][0]["properties"]["arr_field"] == [SINGLE_QUOTE]

            query = EventsQuery(
                after="-24h",
                event="$pageview",
                kind="EventsQuery",
                where=[
                    "has(JSONExtract(ifNull(properties.arr_field,'[]'),'Array(String)'), 'I\"m a string with a \" in it')"
                ],
                orderBy=["timestamp ASC"],
                select=["*"],
            )

            runner = EventsQueryRunner(query=query, team=self.team)
            response = runner.run()
            assert isinstance(response, CachedEventsQueryResponse)
            assert len(response.results) == 1
            assert response.results[0][0]["properties"]["arr_field"] == [DOUBLE_QUOTE]

    @also_test_with_different_timezones
    @snapshot_clickhouse_queries
    def test_absolute_date_range(self):
        self._create_events(
            data=[
                (  # Event two hours BEFORE THE START of the day
                    "p17",
                    "2020-01-11T22:00:00",
                    {},
                ),
                (  # Event one hour after the start of the day
                    "p2",
                    "2020-01-12T01:00:00",
                    {},
                ),
                (  # Event right in the middle of the day
                    "p3",
                    "2020-01-12T12:00:00",
                    {},
                ),
                (  # Event one hour before the end of the day
                    "p1",
                    "2020-01-12T23:00:00",
                    {},
                ),
                (  # Event two hours AFTER THE END of the day
                    "p3",
                    "2020-01-13T02:00:00",
                    {},
                ),
            ]
        )

        flush_persons_and_events()

        query = EventsQuery(
            after="2020-01-12",
            before="2020-01-12T23:59:59",
            event="$pageview",
            kind="EventsQuery",
            orderBy=["timestamp ASC"],
            select=["*"],
        )

        runner = EventsQueryRunner(query=query, team=self.team)

        response = runner.run()
        assert isinstance(response, CachedEventsQueryResponse)
        assert [row[0]["timestamp"] for row in response.results] == [
            datetime(2020, 1, 12, 1, 0, 0, tzinfo=self.team.timezone_info),
            datetime(2020, 1, 12, 12, 0, 0, tzinfo=self.team.timezone_info),
            datetime(2020, 1, 12, 23, 0, 0, tzinfo=self.team.timezone_info),
        ]
