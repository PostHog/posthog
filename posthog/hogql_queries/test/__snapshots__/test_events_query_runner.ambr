# serializer version: 1
# name: TestEventsQueryRunner.test_absolute_date_range
  '''
  SELECT tuple(events.uuid, events.event, events.properties, toTimeZone(events.timestamp, 'UTC'), events.team_id, events.distinct_id, events.elements_chain, toTimeZone(events.created_at, 'UTC'))
  FROM events
  WHERE and(equals(events.team_id, 99999), equals(events.event, '$pageview'), less(toTimeZone(events.timestamp, 'UTC'), toDateTime64('2020-01-12 23:59:59.000000', 6, 'UTC')), greater(toTimeZone(events.timestamp, 'UTC'), toDateTime64('2020-01-12 00:00:00.000000', 6, 'UTC')))
  ORDER BY toTimeZone(events.timestamp, 'UTC') ASC
  LIMIT 101
  OFFSET 0 SETTINGS readonly=2,
                    max_execution_time=60,
                    allow_experimental_object_type=1,
                    format_csv_allow_double_quotes=0,
                    max_ast_elements=4000000,
                    max_expanded_ast_elements=4000000,
                    max_bytes_before_external_group_by=0
  '''
# ---
# name: TestEventsQueryRunner.test_absolute_date_range_minus_utc
  '''
  SELECT tuple(events.uuid, events.event, events.properties, toTimeZone(events.timestamp, 'America/Phoenix'), events.team_id, events.distinct_id, events.elements_chain, toTimeZone(events.created_at, 'America/Phoenix'))
  FROM events
  WHERE and(equals(events.team_id, 99999), equals(events.event, '$pageview'), less(toTimeZone(events.timestamp, 'America/Phoenix'), toDateTime64('2020-01-12 23:59:59.000000', 6, 'America/Phoenix')), greater(toTimeZone(events.timestamp, 'America/Phoenix'), toDateTime64('2020-01-12 00:00:00.000000', 6, 'America/Phoenix')))
  ORDER BY toTimeZone(events.timestamp, 'America/Phoenix') ASC
  LIMIT 101
  OFFSET 0 SETTINGS readonly=2,
                    max_execution_time=60,
                    allow_experimental_object_type=1,
                    format_csv_allow_double_quotes=0,
                    max_ast_elements=4000000,
                    max_expanded_ast_elements=4000000,
                    max_bytes_before_external_group_by=0
  '''
# ---
# name: TestEventsQueryRunner.test_absolute_date_range_plus_utc
  '''
  SELECT tuple(events.uuid, events.event, events.properties, toTimeZone(events.timestamp, 'Asia/Tokyo'), events.team_id, events.distinct_id, events.elements_chain, toTimeZone(events.created_at, 'Asia/Tokyo'))
  FROM events
  WHERE and(equals(events.team_id, 99999), equals(events.event, '$pageview'), less(toTimeZone(events.timestamp, 'Asia/Tokyo'), toDateTime64('2020-01-12 23:59:59.000000', 6, 'Asia/Tokyo')), greater(toTimeZone(events.timestamp, 'Asia/Tokyo'), toDateTime64('2020-01-12 00:00:00.000000', 6, 'Asia/Tokyo')))
  ORDER BY toTimeZone(events.timestamp, 'Asia/Tokyo') ASC
  LIMIT 101
  OFFSET 0 SETTINGS readonly=2,
                    max_execution_time=60,
                    allow_experimental_object_type=1,
                    format_csv_allow_double_quotes=0,
                    max_ast_elements=4000000,
                    max_expanded_ast_elements=4000000,
                    max_bytes_before_external_group_by=0
  '''
# ---
